src/libraries/System.Net.Security/src/System/Net/Security/TlsFrameHelper.cs

```diff
@@ -0,0 +1,239 @@
+                                header.Version = SslProtocols.Tls11;
+                                break;
+                            case 1:
+                                header.Version = SslProtocols.Tls;
+                                break;
+                            case 0:
+#pragma warning disable 0618
+                                header.Version = SslProtocols.Ssl3;
+#pragma warning restore 0618
+                                break;
+                            default:
+                                header.Version = SslProtocols.None;
+                                break;
+                        }
+                    }
+                    else
+                    {
+                        header.Length = -1;
+                        header.Version = SslProtocols.None;
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        // Returns frame size e.g. header + content
+        public static int GetFrameSize(ReadOnlySpan<byte> frame)
+        {
+            if (frame.Length < 5 || frame[1] < 3)
+            {
+                return - 1;
+            }
+
+            return ((frame[3] << 8) | frame[4]) + HeaderSize;
+        }
+
+        public static bool TryGetHandshakeInfo(ReadOnlySpan<byte> frame, ref TlsFrameHandshakeInfo info)
+        {
+            if (frame.Length < 6 || frame[0] != (byte)TlsContentType.Handshake)
+            {
+                return false;
+            }
+
+            // This will not fail since we have enough data.
+            TryGetFrameHeader(frame, ref info.Header);
+            info.SupportedVersions = info.Header.Version;
+
+            info.HandshakeType = (TlsHandshakeType)frame[5];
+
+            if (info.HandshakeType == TlsHandshakeType.ClientHello)
+            {
+                info.TargetName = SniHelper.GetServerName(frame);
+            }
+
+            return true;
+        }
+
+        public static bool TryGetAlertInfo(ReadOnlySpan<byte> frame, ref TlsAlertLevel level, ref TlsAlertDescription description)
+        {
+            if (frame.Length < 7 || frame[0] != (byte)TlsContentType.Alert)
+            {
+                return false;
+            }
+
+            level = (TlsAlertLevel)frame[5];
+            description = (TlsAlertDescription)frame[6];
+
+            return true;
+        }
+
+        private static byte[] CreateProtocolVersionAlert(SslProtocols version)
+        {
+            switch (version)
+            {
+                case SslProtocols.Tls13:
+                    return _protocolMismatch13;
+                case SslProtocols.Tls12:
+                    return _protocolMismatch12;
+                case SslProtocols.Tls11:
+                    return _protocolMismatch11;
+                case SslProtocols.Tls:
+                    return _protocolMismatch10;
+                default:
+                    return Array.Empty<byte>();
+            }
+        }
```

Comment:
Fwiw, this can be written a bit more concisely as:
```C#
private static byte[] CreateProtocolVersionAlert(SslProtocols version) =>
    version switch
    {
        SslProtocols.Tls13 => _protocolMismatch13,
        SslProtocols.Tls12 => _protocolMismatch12,
        SslProtocols.Tls11 => _protocolMismatch11,
        SslProtocols.Tls => _protocolMismatch10,
        _ => Array.Empty<byte>(),
    };
```